File: BE/src/main/java/com/fpt/evcare/service/MessageAssignmentService.java
package com.fpt.evcare.service;

import com.fpt.evcare.dto.request.message.MessageAssignmentRequest;
import com.fpt.evcare.dto.response.MessageAssignmentResponse;
import com.fpt.evcare.dto.response.PageResponse;
import com.fpt.evcare.dto.response.UserResponse;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface MessageAssignmentService {
    
    /**
     * Phân công staff chat với customer
     * - Chỉ admin mới có quyền
     * - 1 customer chỉ được assign 1 staff tại 1 thời điểm
     * - Nếu đã assign, sẽ inactive assignment cũ và tạo mới
     */
    MessageAssignmentResponse assignCustomerToStaff(
        MessageAssignmentRequest request, 
        UUID assignedById
    );
    
    /**
     * Lấy staff được assign cho customer
     */
    MessageAssignmentResponse getAssignmentByCustomerId(UUID customerId);
    
    /**
     * Lấy danh sách customers được assign cho staff
     */
    PageResponse<MessageAssignmentResponse> getAssignmentsByStaffId(UUID staffId, Pageable pageable);
    
    /**
     * Lấy tất cả assignments (cho admin)
     */
    PageResponse<MessageAssignmentResponse> getAllAssignments(Pageable pageable);
    
    /**
     * Lấy danh sách customers chưa được assign
     */
    PageResponse<UserResponse> getUnassignedCustomers(Pageable pageable);
    
    /**
     * Hủy assignment (inactive)
     */
    void deactivateAssignment(UUID assignmentId, UUID adminId);
    
    /**
     * Reassign customer sang staff khác
     */
    MessageAssignmentResponse reassignCustomer(
        UUID customerId, 
        UUID newStaffId, 
        UUID assignedById
    );
    
    /**
     * Tự động phân công staff cho customer (load balancing)
     * - Chọn staff có ít customers nhất
     * - Nếu customer đã có assignment, return existing
     */
    MessageAssignmentResponse autoAssignCustomerToStaff(UUID customerId);
}

File: FE/src/service/messageAssignmentService.ts
import { apiClient } from "./api";
import type { ApiResponse } from "../types/api";
import type {
  MessageAssignmentResponse,
  MessageAssignmentRequest,
  PageResponse,
} from "../types/message.types";
import type { UserResponse } from "../types/admin/user";

const API_BASE = "/message-assignments";

export const messageAssignmentService = {
  /**
   * Phân công customer cho staff (Admin only)
   */
  assignCustomerToStaff: async (
    data: MessageAssignmentRequest,
    adminId: string
  ) => {
    const response = await apiClient.post<
      ApiResponse<MessageAssignmentResponse>
    >(`${API_BASE}`, data, {
      headers: {
        "user-id": adminId,
      },
    });
    return response;
  },

  /**
   * Lấy staff của customer
   */
  getAssignmentByCustomerId: async (customerId: string) => {
    const response = await apiClient.get<
      ApiResponse<MessageAssignmentResponse>
    >(`${API_BASE}/customer/${customerId}`);
    return response;
  },

  /**
   * Lấy customers của staff
   */
  getAssignmentsByStaffId: async (
    staffId: string,
    page: number = 0,
    pageSize: number = 20
  ) => {
    const response = await apiClient.get<
      ApiResponse<PageResponse<MessageAssignmentResponse>>
    >(`${API_BASE}/staff/${staffId}`, {
      params: { page, pageSize },
    });
    return response;
  },

  /**
   * Lấy tất cả assignments (Admin only)
   */
  getAllAssignments: async (page: number = 0, pageSize: number = 20) => {
    const response = await apiClient.get<
      ApiResponse<PageResponse<MessageAssignmentResponse>>
    >(`${API_BASE}`, {
      params: { page, pageSize },
    });
    return response;
  },

  /**
   * Lấy customers chưa được phân công
   */
  getUnassignedCustomers: async (page: number = 0, pageSize: number = 20) => {
    const response = await apiClient.get<
      ApiResponse<PageResponse<UserResponse>>
    >(`${API_BASE}/unassigned-customers`, {
      params: { page, pageSize },
    });
    return response;
  },

  /**
   * Hủy assignment
   */
  deactivateAssignment: async (assignmentId: string, adminId: string) => {
    const response = await apiClient.put<ApiResponse<string>>(
      `${API_BASE}/${assignmentId}/deactivate`,
      {},
      {
        headers: {
          "user-id": adminId,
        },
      }
    );
    return response;
  },

  /**
   * Reassign customer
   */
  reassignCustomer: async (
    customerId: string,
    newStaffId: string,
    adminId: string
  ) => {
    const response = await apiClient.put<
      ApiResponse<MessageAssignmentResponse>
    >(
      `${API_BASE}/reassign`,
      {},
      {
        params: { customerId, newStaffId },
        headers: {
          "user-id": adminId,
        },
      }
    );
    return response;
  },

  /**
   * Tự động phân công staff cho customer (load balancing)
   */
  autoAssignStaff: async (customerId: string) => {
    const response = await apiClient.post<
      ApiResponse<MessageAssignmentResponse>
    >(`${API_BASE}/auto-assign/${customerId}`);
    return response;
  },
};

